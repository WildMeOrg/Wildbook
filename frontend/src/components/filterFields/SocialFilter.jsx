import React from "react";
import FormGroupMultiSelect from "../Form/FormGroupMultiSelect";
import { Form } from "react-bootstrap";
import { FormattedMessage } from "react-intl";
import Description from "../Form/Description";
import AndSelector from "../AndSelector";

export default function SocialFilter({ data, onChange }) {
  const [isUnitChecked, setIsUnitChecked] = React.useState(false);
  const [isRoleChecked, setIsRoleChecked] = React.useState(false);
  const socialRoleOptions =
    data?.relationshipRole?.map((item) => {
      return {
        value: item,
        label: item,
      };
    }) || [];

  const socialUnitOptions =
    data?.socialUnitRole?.map((item) => {
      return {
        value: item,
        label: item,
      };
    }) || [];

  return (
    <div>
      <h3>
        <FormattedMessage id="FILTER_SOCIAL" />
      </h3>
      <Description>
        <FormattedMessage id="FILTER_SOCIAL_DESC" />
      </Description>

      <div className="d-flex flex-row justify-content-between mt-2">
        <Form.Label>
          <FormattedMessage id="FILTER_SOCIAL_UNIT" />
        </Form.Label>

        <Form.Check
          type="checkbox"
          id="custom-checkbox_unit"
          label={<FormattedMessage id="USE_AND_OPERATOR" />}
          checked={isUnitChecked}
          onChange={() => {
            setIsUnitChecked(!isUnitChecked);
          }}
        />
      </div>

      {isUnitChecked ? (
        <AndSelector
          isMulti={true}
          noLabel={true}
          noDesc={true}
          label="FILTER_SOCIAL_UNIT"
          onChange={onChange}
          options={socialUnitOptions}
          field="individualSocialUnits"
          term="terms"
          filterId={"individualSocialUnits"}
          filterKey={"Social Group Unit"}
        />
      ) : (
        <FormGroupMultiSelect
          isMulti={true}
          noLabel={true}
          noDesc={true}
          label="FILTER_SOCIAL_UNIT"
          onChange={onChange}
          options={socialUnitOptions}
          field="individualSocialUnits"
          term="terms"
          filterId={"individualSocialUnits"}
          filterKey={"Social Group Unit"}
        />
      )}

      <div className="d-flex flex-row justify-content-between mt-2">
        <Form.Label>
          <FormattedMessage id="FILTER_RELATIONSHIP_ROLE" />
        </Form.Label>

        <Form.Check
          type="checkbox"
          id="custom-checkbox_role"
          label={<FormattedMessage id="USE_AND_OPERATOR" />}
          checked={isRoleChecked}
          onChange={() => {
            setIsRoleChecked(!isRoleChecked);
          }}
        />
      </div>

      {isRoleChecked ? (
        <AndSelector
          isMulti={true}
          noDesc={true}
          noLabel={true}
          label="FILTER_RELATIONSHIP_ROLE"
          options={socialRoleOptions}
          onChange={onChange}
          field="individualRelationshipRoles"
          term={"terms"}
          filterId={"individualRelationshipRoles"}
          filterKey={"Relationship Role"}
        />
      ) : (
        <FormGroupMultiSelect
          isMulti={true}
          noDesc={true}
          noLabel={true}
          label="FILTER_RELATIONSHIP_ROLE"
          options={socialRoleOptions}
          onChange={onChange}
          field="individualRelationshipRoles"
          term={"terms"}
          filterId={"individualRelationshipRoles"}
          filterKey={"Relationship Role"}
        />
      )}
    </div>
  );
}
