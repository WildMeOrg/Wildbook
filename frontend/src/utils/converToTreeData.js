const convertToTreeData = (locationData) => {
  if (!Array.isArray(locationData)) return [];
  return locationData.map((location) => ({
    title: location.id,
    value: location.id,
    children:
      location.locationID?.length > 0
        ? convertToTreeData(location.locationID)
        : [],
  }));
};

export default convertToTreeData;
