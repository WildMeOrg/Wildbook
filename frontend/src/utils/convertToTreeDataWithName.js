const convertToTreeDataWithName = (locationData) => {
  if (!Array.isArray(locationData)) return [];
  return locationData.map((location) => ({
    title: location.name,
    value: location.id,
    children:
      location.locationID?.length > 0
        ? convertToTreeDataWithName(location.locationID)
        : [],
  }));
};

export default convertToTreeDataWithName;
