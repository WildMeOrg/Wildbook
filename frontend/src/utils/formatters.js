import { format } from "date-fns";

export const formatDate = (input, fancy = false, fallback = "") => {
  const formatter = fancy ? "PP" : "yyyy-MM-dd HH:mm";
  try {
    const jsDate = typeof input === "string" ? new Date(input) : input;
    const formattedDate = format(jsDate, formatter);
    return formattedDate;
  } catch (_error) {
    // console.error(error);
    return fallback;
  }
};
