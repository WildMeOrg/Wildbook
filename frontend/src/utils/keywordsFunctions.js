import axios from "axios";

export async function postRestKeyword(payload, opts = {}) {
  const res = await axios.post(`/RestKeyword`, payload, {
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
  });
  return res.data; 
}

export async function getAssetKeywords(mid) {
  const res = await axios.get(`/RestKeyword`, {
    params: { assetIds: String(mid) },   
    withCredentials: true,
    headers: { Accept: "application/json" },
  });
  const data = res.data || {};
  const byMid = (data.results && data.results[mid]) || {};
  return Object.entries(byMid).map(([id, name]) => ({ id, name }));
}
export function addExistingKeyword(mid, keywordId) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], add: [keywordId] } });
}

export function addNewKeywordText(mid, text) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], newAdd: [text] } });
}

export function removeKeyword(mid, keywordId) {
  return postRestKeyword({ onMediaAssets: { assetIds: [mid], remove: [keywordId] } });
}
