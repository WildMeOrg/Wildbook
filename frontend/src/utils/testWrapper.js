import React from "react";
import { IntlProvider } from "react-intl";
import messages from "../locale/en.json";

export const wrapper = ({ children }) => (
  <IntlProvider locale="en" messages={messages}>
    {children}
  </IntlProvider>
);
