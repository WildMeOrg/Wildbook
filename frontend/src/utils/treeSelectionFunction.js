export function findNodeByValue(treeData, value) {
  for (const node of treeData || []) {
    if (node?.value === value) return node;
    if (node?.children?.length) {
      const found = findNodeByValue(node.children, value);
      if (found) return found;
    }
  }
  return null;
}

export function getAllDescendantValues(node) {
  const res = [];
  (node?.children || []).forEach((child) => {
    if (child?.value != null) res.push(child.value);
    res.push(...getAllDescendantValues(child));
  });
  return res;
}

export function expandIds(locationOptions, values = []) {
  const set = new Set();
  (values || []).forEach((val) => {
    if (!val) return;
    set.add(val);
    const node = findNodeByValue(locationOptions, val); 
    if (node) {
      getAllDescendantValues(node).forEach((cid) => {
        if (cid != null) set.add(cid);
      });
    }
  });
  return Array.from(set);
}