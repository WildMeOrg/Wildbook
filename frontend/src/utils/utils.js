import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios';
import { QueryClient, QueryClientProvider } from 'react-query';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import messages from '../locale/en.json';
import AuthContext from '../AuthProvider';

jest.mock('axios');

const renderWithProviders = (ui, isLoggedIn = true) => {
  const queryClient = new QueryClient();
  return render(
    <QueryClientProvider client={queryClient}>
      <IntlProvider locale="en" messages={messages}>
      <AuthContext.Provider value={{ isLoggedIn }}>
        <MemoryRouter>{ui}</MemoryRouter>
      </AuthContext.Provider>
      </IntlProvider>
    </QueryClientProvider>
  );
};

const renderWithRouter = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);
  return render(ui, { wrapper: BrowserRouter });
};

const fireInput = (placeholder, value) => {
  const input = screen.getByPlaceholderText(placeholder);
  fireEvent.change(input, { target: { value } });
  return input;
};

const clickButton = (buttonText) => {
  fireEvent.click(screen.getByRole('button', { name: buttonText }));
};

const waitForText = async (text) => {
  await waitFor(() => expect(screen.getByText(text)).toBeInTheDocument());
};

const mockAxiosSuccess = (mockedData) => {
  axios.get.mockResolvedValue({ data: mockedData });
};

const mockAxiosFailure = () => {
  axios.get.mockRejectedValue(new Error('Network error'));
};

const mockComponent = (testId) => {
  return () => {
    const React = require("react");
    const MockLandingImage = () =>
      React.createElement("div", { "data-testid": testId });
    MockLandingImage.displayName = testId;
    return MockLandingImage;
  }
}

export * from '@testing-library/react';
export {
  renderWithRouter as render,
  fireInput,
  clickButton,
  waitForText,
  mockAxiosSuccess,
  mockAxiosFailure,
  renderWithProviders,
  mockComponent,
};
