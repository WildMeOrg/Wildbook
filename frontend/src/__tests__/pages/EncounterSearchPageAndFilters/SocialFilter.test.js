import React from "react";
import { screen, fireEvent } from "@testing-library/react";
import SocialFilter from "../../../components/filterFields/SocialFilter";
import { renderWithProviders } from "../../../utils/utils";

const mockStore = {
  formFilters: [],
  addFilter: jest.fn(),
  removeFilter: jest.fn(),
  removeFilterByFilterKey: jest.fn(),
};

const mockData = {
  relationshipRole: ["Parent", "Sibling"],
  socialUnitRole: ["Family", "Group"],
};

describe("SocialFilter Component", () => {
  beforeEach(() => {
    mockStore.formFilters = [];
    jest.clearAllMocks();
  });

  const renderComponent = () => {
    renderWithProviders(<SocialFilter data={mockData} store={mockStore} />);
  };

  test("renders labels and descriptions correctly", () => {
    renderComponent();

    expect(screen.getByText("FILTER_SOCIAL")).toBeInTheDocument();
    expect(screen.getByText("FILTER_SOCIAL_DESC")).toBeInTheDocument();
    expect(screen.getByText("FILTER_GROUP_BEHAVIOR")).toBeInTheDocument();
    expect(screen.getByText("FILTER_GROUP_COMPOSITION")).toBeInTheDocument();
    expect(screen.getByText("FILTER_SOCIAL_UNIT")).toBeInTheDocument();
    expect(screen.getByText("FILTER_RELATIONSHIP_ROLE")).toBeInTheDocument();
  });

  test("handles 'AND' operator checkbox for Social Unit correctly", () => {
    renderComponent();

    const socialUnitAndCheckbox =
      screen.getAllByLabelText("USE_AND_OPERATOR")[0];

    fireEvent.click(socialUnitAndCheckbox);

    expect(mockStore.removeFilterByFilterKey).not.toHaveBeenCalled();

    fireEvent.click(socialUnitAndCheckbox);

    expect(mockStore.removeFilterByFilterKey).toHaveBeenCalledWith(
      "Social Group Unit",
    );
  });

  test("handles 'AND' operator checkbox for Relationship Role correctly", () => {
    renderComponent();

    const relationshipRoleAndCheckbox = screen.getAllByRole("checkbox", {
      name: /USE_AND_OPERATOR/i,
    })[1];

    fireEvent.click(relationshipRoleAndCheckbox);

    expect(mockStore.removeFilterByFilterKey).not.toHaveBeenCalled();

    fireEvent.click(relationshipRoleAndCheckbox);

    expect(mockStore.removeFilterByFilterKey).toHaveBeenCalledWith(
      "Relationship Role",
    );
  });
});
