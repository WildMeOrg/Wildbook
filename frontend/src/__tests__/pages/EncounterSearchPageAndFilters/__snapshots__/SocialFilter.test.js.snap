// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialFilter Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      FILTER_SOCIAL
    </h4>
    <div
      style="font-size: 14px; font-weight: 400; line-height: 21px; color: rgb(175, 179, 183); margin-bottom: 1rem;"
    >
      FILTER_SOCIAL_DESC
    </div>
    <div
      class="mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_GROUP_BEHAVIOR
      </label>
      <input
        class="form-control"
        placeholder="Type here"
        type="text"
        value=""
      />
    </div>
    <div
      class="mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_GROUP_COMPOSITION
      </label>
      <input
        class="form-control"
        placeholder="Type here"
        type="text"
        value=""
      />
    </div>
    <div
      class="d-flex flex-row justify-content-between mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_SOCIAL_UNIT
      </label>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="custom-checkbox_unit"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="custom-checkbox_unit"
          title=""
        >
          USE_AND_OPERATOR
        </label>
      </div>
    </div>
    <div
      class="mt-2"
    >
      <div
        class="basic-multi-select css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-18-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="select__control css-48d7p1-control"
        >
          <div
            class="select__value-container select__value-container--is-multi css-hlgwow"
          >
            <div
              class="select__placeholder css-1jqq78o-placeholder"
              id="react-select-18-placeholder"
            >
              Select one or more
            </div>
            <div
              class="select__input-container css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-18-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="select__input"
                id="react-select-18-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="select__indicators css-1wy0on6"
          >
            <span
              class="select__indicator-separator css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="select__indicator select__dropdown-indicator css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="d-flex flex-row justify-content-between mt-2"
    >
      <label
        class="form-label"
      >
        FILTER_RELATIONSHIP_ROLE
      </label>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="custom-checkbox_role"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="custom-checkbox_role"
          title=""
        >
          USE_AND_OPERATOR
        </label>
      </div>
    </div>
    <div
      class="mt-2"
    >
      <div
        class="basic-multi-select css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-19-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="select__control css-48d7p1-control"
        >
          <div
            class="select__value-container select__value-container--is-multi css-hlgwow"
          >
            <div
              class="select__placeholder css-1jqq78o-placeholder"
              id="react-select-19-placeholder"
            >
              Select one or more
            </div>
            <div
              class="select__input-container css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-19-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="select__input"
                id="react-select-19-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="select__indicators css-1wy0on6"
          >
            <span
              class="select__indicator-separator css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="select__indicator select__dropdown-indicator css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
